llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -4712,23 +4733,29 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
     if (match(CmpLHS, m_Or(m_Value(X), m_Value(Y))) &&
         match(CmpRHS, m_Zero())) {
       // (X | Y) == 0 implies X == 0 and Y == 0.
-      if (Value *V = simplifySelectWithEquivalence(X, CmpRHS, TrueVal, FalseVal,
-                                                   Q, MaxRecurse))
+      if (Value *V = simplifySelectWithEquivalence(
+              {{X, CmpRHS}, {Y, CmpRHS}}, TrueVal, FalseVal, Q, MaxRecurse))
+        return V;
+      if (Value *V = simplifySelectWithEquivalence({{X, CmpRHS}}, TrueVal,
+                                                   FalseVal, Q, MaxRecurse))
         return V;
-      if (Value *V = simplifySelectWithEquivalence(Y, CmpRHS, TrueVal, FalseVal,
-                                                   Q, MaxRecurse))
+      if (Value *V = simplifySelectWithEquivalence({{Y, CmpRHS}}, TrueVal,
+                                                   FalseVal, Q, MaxRecurse))
```

Comment:
Why do we need to do the replacement of both variables, and then also try to replace only a single variable?

Reply:
Because our simplification goal is TrueVal == FalseVal, not the complete simplification of either arm independently. If we only do the double replace we can hit some cases where we oversimplify a side and then fail to optimize our the select.

Reply:
Do you have an example where this happens? I can see how this could occur without https://github.com/llvm/llvm-project/pull/121753, but with it, I'd expect complete simplification to always be better than the individual simplifications.

Reply:
Yeah I think your patch fixes it. I will rebase.

Reply:
Rebased + dropped extra calls.